### PR TITLE
feat: improve topic creation api

### PR DESCRIPTION
**Pull request by fraidev**
Simple (1 partition and 1 replica):
```python
fluvio_admin = FluvioAdmin.connect()
fluvio_admin.create_topic("topic_name")
```

or

```python
fluvio_admin = FluvioAdmin.connect()
topic_spec = TopicSpec.new()
fluvio_admin.create_topic("topic_name", topic_spec)
```

Advanced:
```python
fluvio_admin = FluvioAdmin.connect()
topic_spec = (
    TopicSpec.create()
    .with_partitions(3)
    .with_replications(3)
    .with_max_partition_size("1Gb")
    .with_retention_time(3600)
    .with_segment_size("10M")
    .as_system_topic()
    .build()
)
fluvio_admin.create_topic("topic_name", topic_spec)
```

~I am using `NewTopic` only because [kafka-python](https://github.com/dpkp/kafka-python) also uses a class called `NewTopic`~

### CODE REVIEW

File: fluvio/new_topic.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
+from dataclasses import dataclass, replace
+from typing import Optional, List, Union
+from enum import Enum
+from .specs import TopicSpec, PartitionMap
+from .utils import parse_byte_size
+from humanfriendly import parse_timespan
+
+
+class TopicMode(Enum):
```
Comment (sehz) [inline fluvio/new_topic.py:9]: Instead of creating python enum, should use derive macro like here: https://pyo3.rs/v0.23.2/conversions/traits.html

**Comment by fraidev** [comment]
> ```python
> NewTopic.create
> ```

Sure, but should I merge them or rename the actual `TopicSpec`?

Because the actual `TopicSpec` is a wrapper of the inner `TopicSpec` from Rust.

The `NewTopic` that I created is the builder of `TopicSpec`. Allowing specifying the number of the partitions and replications with these methods, instead just call `new_computed` method.

Merge them would be something like this, what do you think?

```python
fluvio_admin = FluvioAdmin.connect()
topic_spec = (
    TopicSpec.new_computed(partitions=3, replications=3, ignore=True)
    .with_max_partition_size("1Gb")
    .with_retention_time(3600)
    .with_segment_size("10M")
    .as_system_topic()
    .build()
)
fluvio_admin.create_topic(self.topic, topic_spec)
```

Another alternative is rename `NewTopic` to `TopicSpecBuilder` or rename the `NewTopic` to `TopicSpec` and `TopicSpec` to something else.

```python
fluvio_admin = FluvioAdmin.connect()
topic_spec = (
    TopicSpecBuilder.create()
    .with_partitions(3)
    .with_replications(3)
    .with_max_partition_size("1Gb")
    .with_retention_time(3600)
    .with_segment_size("10M")
    .as_system_topic()
    .build()
)
fluvio_admin.create_topic(self.topic, topic_spec)
```

**Comment by digikata** [comment]
The first one is good w/ TopicSpec.   If you can do it with the direct annotations in lib.rs PyO3 or appropriate module use that. There are a lot of instances in the fluvio python client that rename some FluvioStruct to _FluvioStruct internally, so a pythonic FluvioStruct can be supplied to the interpreter too, but it's extra layers that should have some interface value if thats done.  (there are legacy portions of the code that does it unecessarily)

File: fluvio/new_topic.py (hunk 1 above)
Comment (fraidev) [reply fluvio/new_topic.py:9]: This new enum is only for the builder of TopicSpec, if we consider the solution without the build it's not even needed.

Comment (fraidev) [reply fluvio/new_topic.py:9]: Maybe the Compression Enum is worth, thought

**Comment by digikata** [comment]
Oh we shoudl also remove the required partitions, replications, and bool arguments and just support `TopicSpec.new()` with defaults that can be changed w/ the builder style calls as you did with the other parameters

**Comment by fraidev** [comment]
> Oh we shoudl also remove the required partitions, replications, and bool arguments and just support `TopicSpec.new()` with defaults that can be changed w/ the builder style calls as you did with the other parameters

I think that would need a larger refectory on fluvio side. IIRC, we don't have the feature of adding replicas to an already existent a Topic. 

In the builder it was easier because it's calling `new` only after the `build` function.

**Comment by digikata** [comment]
This is only more featured  topic create for the python client, if the topic already exists the create should return an error.

File: src/lib.rs [hunk 2]
```diff
@@ -1111,6 +1113,62 @@ impl TopicSpec {
...
+        } else if compression == "snappy" {
+            self.inner
+                .set_compression_type(CompressionAlgorithm::Snappy);
+        } else if compression == "lz4" {
+            self.inner.set_compression_type(CompressionAlgorithm::Lz4);
+        } else if compression == "zstd" {
+            self.inner.set_compression_type(CompressionAlgorithm::Zstd);
+        }
+        self.inner.set_compression_type(CompressionAlgorithm::Any);
+    }
```
Comment (digikata) [inline src/lib.rs:1171]: This should probably return an error if not recognized, or it could accept a compression type enum

Comment (fraidev) [reply src/lib.rs:1174]: I changed to return result, and an error for this case.
The support for enums that are not tag unions does not seem great in pyo3, it's also not automatically typed, so no benefits at all.

**Review by sehz — APPROVED** [review]

**Review by digikata — APPROVED** [review]

**Pull request merged**